cranelift/egraph/src/egraph.rs

```diff
@@ -0,0 +1,516 @@
+//! recanonicalizing of eclass arguments as they union; but also this
+//! is not usually necessary, because args will have already been
+//! processed and eagerly rewritten as well. In other words, eager
+//! rewriting and the immutable data structure mutually allow each
+//! other to be practical; both work together.
+//!
+//! [^1]: M Willsey, C Nandi, Y R Wang, O Flatt, Z Tatlock, P
+//!       Panchekha. "egg: Fast and Flexible Equality Saturation." In
+//!       POPL 2021. https://dl.acm.org/doi/10.1145/3434304
+
+use crate::ctxhash::{CtxEq, CtxHash, CtxHashMap, Entry};
+use crate::unionfind::UnionFind;
+use crate::{Id, Language};
+use cranelift_entity::PrimaryMap;
+use smallvec::{smallvec, SmallVec};
+use std::marker::PhantomData;
+
+/// An egraph.
+pub struct EGraph<L: Language> {
+    /// Node-allocation arena.
+    pub nodes: Vec<L::Node>,
+    /// Hash-consing map from Nodes to eclass IDs.
+    node_map: CtxHashMap<NodeKey, Id>,
+    /// Eclass definitions. Each eclass consists of an enode, and
+    /// parent pointer to the rest of the eclass.
+    pub classes: PrimaryMap<Id, EClass>,
+    /// Union-find for canonical ID generation. This lets us name an
+    /// eclass with a canonical ID that is the same for all
+    /// generations of the class.
+    pub unionfind: UnionFind,
+}
+
+/// A reference to a node.
+#[derive(Clone, Copy, Debug)]
+pub struct NodeKey {
+    index: u32,
+}
+
+impl NodeKey {
+    fn from_node_idx(node_idx: usize) -> NodeKey {
+        NodeKey {
+            index: u32::try_from(node_idx).unwrap(),
+        }
+    }
+
+    /// Get the node for this NodeKey, given the `nodes` from the
+    /// appropriate `EGraph`.
+    pub fn node<'a, L: Language>(&self, nodes: &'a [L::Node]) -> &'a L::Node {
+        &nodes[self.index as usize]
+    }
+
+    fn bits(self) -> u32 {
+        self.index
+    }
+
+    fn from_bits(bits: u32) -> Self {
+        NodeKey { index: bits }
+    }
+}
+
+struct NodeKeyCtx<'a, L: Language> {
+    nodes: &'a [L::Node],
+    node_ctx: &'a L,
+}
+
+impl<'ctx, L: Language> CtxEq<NodeKey, NodeKey> for NodeKeyCtx<'ctx, L> {
+    fn ctx_eq(&self, a: &NodeKey, b: &NodeKey, uf: &mut UnionFind) -> bool {
+        let a = a.node::<L>(self.nodes);
+        let b = b.node::<L>(self.nodes);
+        self.node_ctx.ctx_eq(a, b, uf)
+    }
+}
+
+impl<'ctx, L: Language> CtxHash<NodeKey> for NodeKeyCtx<'ctx, L> {
+    fn ctx_hash(&self, value: &NodeKey, uf: &mut UnionFind) -> u64 {
+        self.node_ctx.ctx_hash(value.node::<L>(self.nodes), uf)
+    }
+}
+
+/// An EClass entry. Contains either a single new enode and a parent
+/// eclass (i.e., adds one new enode), or unions two parent eclasses
+/// together.
+#[derive(Debug, Clone, Copy)]
+pub struct EClass {
+    // formats:
+    //
+    // 00 | unused  (31 bits)         | NodeKey (31 bits)
+    // 01 | eclass_parent   (31 bits) | NodeKey (31 bits)
+    // 10 | eclass_parent_1 (31 bits) | eclass_parent_id_2 (31 bits)
+    bits: u64,
+}
```

Comment:
If I'm reading this correctly, then an e-class containing `n` nodes will have `O(n)` "pointer" chases to enumerate its e-nodes? I suppose that is fine for small e-classes, and maybe most are small, but we might want to investigate those e-node counts in the future and maybe have a variant for when there are a bunch of e-nodes in an e-class that stores them together in a single contiguous region of memory. Or use HAMTs or something.

Reply:
Yep, that's a reasonable optimization, I agree. In practice it didn't seem to matter for the benchmarks I was looking at but it's something to keep an eye on in the future.